### PR TITLE
[RISCV][ISel] Remove redundant min/max in saturating truncation

### LINKED ISSUE #73424: [RISCV] Vector saturating truncation not using VNCLIP
Closely related to #68466 and #58266, but for a different platform.

Much like x86's (V)PACK(U,S)S and ARM's (U,S)QXTN(2), RISC-V's V extension includes VNCLIP(U).W(V,X,I), which performs a right shift and then performs either signed or unsigned saturation to vector of integers with half the width. With a shift value of 0 (whether from an immediate or the register; the V extension specification doesn't specifically talk about this use case), the instruction can be used to perform a saturating truncation from vectors of one integer type, the the type with half the width.

Given the complexity of matching saturating truncation, and the apparent duplication between targets of this functionality, it could be worth considering matching it earlier in the pipeline. Matching a saturating truncation preceded by a shift should be much simpler and can be done with TableGen.

Example code with current code gen, next to x86 and AArch64 equivalents.

https://rust.godbolt.org/z/bKdsq8GGM

### PR DESCRIPTION
**Pull request by sun-jacobi**
This patch closed #73424, which is also a missed-optimization case similar to #68466 on X86.

## Source Code
```
define void @trunc_sat_i8i16(ptr %x, ptr %y) {
  %1 = load <8 x i16>, ptr %x, align 16
  %2 = tail call <8 x i16> @llvm.smax.v8i16(<8 x i16> %1, <8 x i16> <i16 -128, i16 -128, i16 -128, i16 -128, i16 -128, i16 -128, i16 -128, i16 -128>)
  %3 = tail call <8 x i16> @llvm.smin.v8i16(<8 x i16> %2, <8 x i16> <i16 127, i16 127, i16 127, i16 127, i16 127, i16 127, i16 127, i16 127>)
  %4 = trunc <8 x i16> %3 to <8 x i8>
  store <8 x i8> %4, ptr %y, align 8
  ret void
}
```
## Before this patch: 
```
trunc_sat_i8i16:                  # @trunc_maxmin_id_i8i16
        vsetivli        zero, 8, e16, m1, ta, ma
        vle16.v v8, (a0)
        li      a0, -128
        vmax.vx v8, v8, a0
        li      a0, 127
        vmin.vx v8, v8, a0
        vsetvli zero, zero, e8, mf2, ta, ma
        vnsrl.wi        v8, v8, 0
        vse8.v  v8, (a1)
        ret
```

## After this patch: 
```
trunc_sat_i8i16:                  # @trunc_maxmin_id_i8i16
	vsetivli	zero, 8, e8, mf2, ta, ma
	vle16.v	v8, (a0)
	csrwi	vxrm, 0
	vnclip.wi	v8, v8, 0
	vse8.v	v8, (a1)
	ret
```

### CODE REVIEW

**Comment by topperc** [comment]
How can you use just vnsrl? vnsrl discards the upper bits without any max or min.

**Comment by sun-jacobi** [comment]
> How can you use just vnsrl? vnsrl discards the upper bits without any max or min.

Yes, you are right. 

But if the range created by a pair of max/min operation is precisely the range of the truncation destination, 
then the pair could be removed.

In the below case, `[-128, 127]` is the value range of u8. 
Since we truncate the u16 vector into a u8 vector, removing the max/min won't change the final result, AFAIU.

**Comment by topperc** [comment]
> > How can you use just vnsrl? vnsrl discards the upper bits without any max or min.
> 
> 
> 
> Yes, you are right. 
> 
> 
> 
> But if the range created by a pair of max/min operation is precisely the range of the truncation destination, 
> 
> then the pair could be removed.
> 
> 
> 
> In the above case, `[-128, 127]` is the value range of u8. 
> 
> Since we truncate the u16 vector into a u8 vector, removing the max/min won't change the final result, AFAIU. 

Let's imagine the input is -32767(0x8001) in i16. The smax with -128 should give -128(0xff80). Then it will be truncated to 0x80 which is -128 in i8.

A vnsrl by itself will drop the upper bits and give 0x01. This is incorrect.

**Review by topperc — CHANGES_REQUESTED** [review]
I don't think this is correct. We need to use vnclip.

**Comment by sun-jacobi** [comment]
Thank you for correcting me. I will fix it soon.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td [hunk 1]
```diff
@@ -2328,6 +2328,63 @@ defm : VPatBinaryVL_VV_VX_VI<riscv_uaddsat_vl, "PseudoVSADDU">;
...
+}
+
+multiclass VPatTruncSatClip<VTypeInfo vti, VTypeInfo wti> {
+  defvar sew = vti.SEW;
+  defvar uminval = !sub(!shl(1, sew), 1);
+  defvar umaxval = 0;
+  defvar sminval = !sub(!shl(1, !sub(sew, 1)), 1);
+  defvar smaxval = !sub(0, !shl(1, !sub(sew, 1)));
+
+  defm : VPatTruncSatClipMaxMin<"PseudoVNCLIP", vti, wti, riscv_umax_vl,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td:2377]: Shouldn't umax/umin use VNCLIPU and smax/smin use VNCLIP?

File: llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td [hunk 2]
```diff
@@ -2328,6 +2328,63 @@ defm : VPatBinaryVL_VV_VX_VI<riscv_uaddsat_vl, "PseudoVSADDU">;
...
+        (wti.Vector undef), (wti.Mask V0), VLOpFrag)),
+    (vti.Mask V0), VLOpFrag)), (!cast<Instruction>(inst#"_WI_"#vti.LMul.MX#"_MASK")
+      (vti.Vector (IMPLICIT_DEF)), wti.RegClass:$rs1, 0,
+      (vti.Mask V0), 0, GPR:$vl, vti.Log2SEW, TA_MA)>;
+}
+
+multiclass VPatTruncSatClipUMin<VTypeInfo vti, VTypeInfo wti, int sminval> {
+  let Predicates = !listconcat(GetVTypePredicates<vti>.Predicates,
+                               GetVTypePredicates<wti>.Predicates) in
+  def : Pat<(vti.Vector (riscv_trunc_vector_vl
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td:2352]: I don't understand this pattern? I don't think we can make a VNCLIP from smin+trunc.

Comment (sun-jacobi) [reply llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td:2352]: Sorry, this pattern is for umin + trunc. I made a mistake.

File: llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td [hunk 3]
```diff
@@ -2328,6 +2328,64 @@ defm : VPatBinaryVL_VV_VX_VI<riscv_uaddsat_vl, "PseudoVSADDU">;
...
+
+multiclass VPatTruncSatClip<VTypeInfo vti, VTypeInfo wti> {
+  defvar sew = vti.SEW;
+  defvar uminval = !sub(!shl(1, sew), 1);
+  defvar umaxval = 0;
+  defvar sminval = !sub(!shl(1, !sub(sew, 1)), 1);
+  defvar smaxval = !sub(0, !shl(1, !sub(sew, 1)));
+
+  defm : VPatTruncSatClipMaxMin<"PseudoVNCLIPU", vti, wti, riscv_umax_vl,
+                                                umaxval, riscv_umin_vl, uminval>;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td:2379]: this is indented too far

File: llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td [hunk 4]
```diff
@@ -2328,6 +2328,64 @@ defm : VPatBinaryVL_VV_VX_VI<riscv_uaddsat_vl, "PseudoVSADDU">;
...
+  defm : VPatTruncSatClipMaxMin<"PseudoVNCLIP", vti, wti, riscv_smin_vl,
+                                                sminval, riscv_smax_vl, smaxval>;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td:2381]: same here

File: llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td [hunk 5]
```diff
@@ -2328,6 +2328,64 @@ defm : VPatBinaryVL_VV_VX_VI<riscv_uaddsat_vl, "PseudoVSADDU">;
...
+}
+
+multiclass VPatTruncSatClip<VTypeInfo vti, VTypeInfo wti> {
+  defvar sew = vti.SEW;
+  defvar uminval = !sub(!shl(1, sew), 1);
+  defvar umaxval = 0;
+  defvar sminval = !sub(!shl(1, !sub(sew, 1)), 1);
+  defvar smaxval = !sub(0, !shl(1, !sub(sew, 1)));
+
+  defm : VPatTruncSatClipMaxMin<"PseudoVNCLIPU", vti, wti, riscv_umax_vl,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td:2378]: Is this unsigned pattern needed? Won't a umax with 0 be deleted by DAG combiner? Leaving only the umin.

**Review by topperc — APPROVED** [review]
LGTM

**Comment by sun-jacobi** [comment]
@topperc 

Could you help me request more reviews?  
It seems that I can not request review (or merge) without the write access.

**Pull request merged**